[Docs] New Drivers and ORMs Section (#10878)
yugabyte/yugabyte-db: docs/content/latest/reference/drivers/java/postgres-jdbc-reference.md

menu:
  latest:
    name: JDBC Drivers
    identifier: ref-postgres-jdbc-driver
    parent: drivers
    weight: 600
isTocNested: true
showAsideToc: true
---

<ul class="nav nav-tabs-alt nav-tabs-yb">

  <li >
    <a href="/latest/reference/drivers/java/yugabyte-jdbc-reference/" class="nav-link">
      <i class="icon-java-bold" aria-hidden="true"></i>
      YugabyteDB JDBC Driver
    </a>
  </li>

  <li >
    <a href="/latest/reference/drivers/java/postgres-jdbc-reference/" class="nav-link active">
      <i class="icon-postgres" aria-hidden="true"></i>
      PostgreSQL JDBC Driver
    </a>
  </li>

</ul>

The [PostgreSQL JDBC driver](https://jdbc.postgresql.org/) is the official JDBC driver for PostgreSQL which can used for connecting to YugabyteDB YSQL. YugabyteDB YSQL has full compatiblity with PostgreSQL JDBC Driver, allows Java programmers to connect to YugabyteDB database to execute DMLs and DDLs using the JDBC APIs.

## Quick Start

Learn how to establish a connection to YugabyteDB database and begin simple CRUD operations using the steps in [Build an Application](/latest/quick-start/build-apps/java/ysql-jdbc) in the Quick Start section.

## Download the Driver Dependency

> Postgres JDBC Drivers are available as maven dependency, you can download the driver by adding the following dependency into the java project.

Reviewer: ```suggestion
Postgres JDBC Drivers are available as a Maven dependency, and you can download the driver by adding the following dependency into the Java project.
```

### Maven Depedency

```xml
<!-- https://mvnrepository.com/artifact/org.postgresql/postgresql -->
<dependency>
  <groupId>org.postgresql</groupId>
  <artifactId>postgresql</artifactId>
  <version>42.2.14</version>
</dependency>
```

### Gradle Dependency

```java
// https://mvnrepository.com/artifact/org.postgresql/postgresql
implementation 'org.postgresql:postgresql:42.2.14'
```

## Fundamentals

> Learn how to perform the common tasks required for Java App development using the PostgreSQL JDBC driver

Reviewer: ```suggestion
Learn how to perform common tasks required for Java application development using the PostgreSQL JDBC driver.
```

<!-- * [Connect to YugabyteDB Database](postgres-jdbc-fundamentals/#connect-to-yugabytedb-database)
* [Configure SSL/TLS](postgres-jdbc-fundamentals/#configure-ssl-tls)
* [Create Table](/postgres-jdbc-fundamentals/#create-table)
* [Read and Write Queries](/postgres-jdbc-fundamentals/#read-and-write-queries) -->

### Connect to YugabyteDB Database

> Java Apps can connect to and query the YugabyteDB database using the `java.sql.DriverManager` class. All the JDBC interfaces required for working with YugabyteDB database will be part of `java.sql.*` package.

Reviewer: ```suggestion
Java applications can connect to and query the YugabyteDB database using the `java.sql.DriverManager` class. The `java.sql.*` package includes all the JDBC interfaces required for working with YugabyteDB.
```

> Use the `DriverManager.getConnection` method for getting connection object for the YugabyteDB Database which can be used for performing DDLs and DMLs against the database.

Reviewer: ```suggestion
Use the `DriverManager.getConnection` method to create a connection object to perform DDLs and DMLs against the database.
```

JDBC Connection String

```java
jdbc://postgresql://hostname:port/database
```

Example JDBC URL for connecting to YugabyteDB can be seen below.

```java
Connection conn = DriverManager.getConnection("jdbc:postgresql://localhost:5433/yugabyte","yugabyte", "yugabyte");
```

| JDBC Params | Description | Default |
| :---------- | :---------- | :------ |
| hostname  | hostname of the YugabyteDB instance | localhost
| port |  Listen port for YSQL | 5433
| database | database name | yugabyte
| user | user for connecting to the database | yugabyte
| password | password for connecting to the database | yugabyte

### Create Table

> Tables on YugabyteDB database can be created using the `java.sql.Statement` interface which is used for executing the `CREATE TABLE` DDL statement.

Reviewer: ```suggestion
Create database tables using the `java.sql.Statement` interface, which is used to execute the `CREATE TABLE` DDL statement.
```

For example

```sql
CREATE TABLE IF NOT EXISTS employee (id int primary key, name varchar, age int, language text)
```

```java
Connection conn = DriverManager.getConnection("jdbc:postgresql://localhost:5433/yugabyte","yugabyte", "yugabyte");
Statment stmt = conn.createStatement();
try {
  stm.execute("CREATE TABLE IF NOT EXISTS employee" +
                    "  (id int primary key, name varchar, age int, language text)");

} catch (SQLException e) {
  System.err.println(e.getMessage());
}
```

> `java.sql.Statement` throws `java.sql.SQLException` exception which needs to handled in the java code. Read more on designing [Database schemas and tables](../../../../explore/ysql-language-features/databases-schemas-tables/).

Reviewer: ```suggestion
`java.sql.Statement` throws the `java.sql.SQLException` exception, which needs to handled in the Java code. Read more on designing [Database schemas and tables](../../../../explore/ysql-language-features/databases-schemas-tables/).
```

### Read and Write Data

#### Insert Data

> In order to write data into YugabyteDB, execute the `INSERT` statement using `java.sql.Statement` interface.

Reviewer: ```suggestion
To write data to YugabyteDB, execute the `INSERT` statement using the `java.sql.Statement` interface.
```

For example

```java
INSERT INTO employee VALUES (1, 'John', 35, 'Java')
```

```java
Connection conn = DriverManager.getConnection("jdbc:postgresql://localhost:5433/yugabyte","yugabyte", "yugabyte");
Statment stmt = conn.createStatement();
try {
  String insertStr = "INSERT INTO employee VALUES (1, 'John', 35, 'Java')";
  stmt.execute(insertStr);

} catch (SQLException e) {
  System.err.println(e.getMessage());
}
```

> For inserting data using JDBC clients, it is always a good pratice to use `java.sql.PreparedStatemet` for executing `INSERT` statements.

Reviewer: ```suggestion
When inserting data using JDBC clients, it is good practice to use `java.sql.PreparedStatemet` for executing `INSERT` statements.
```

```java
Connection conn = DriverManager.getConnection("jdbc:postgresql://localhost:5433/yugabyte","yugabyte", "yugabyte");
Statment stmt = conn.createStatement();
try {

  PreparedStatement pstmt = connection.prepareStatement("INSERT INTO employees (id, name, age, langugage) VALUES (?, ?, ?, ?)");
  pstmt.setInt(1, 1);
  pstmt.setString(2, "John");
  pstmt.setInt(3, 35);
  pstmt.setString(4, "Java");
  pstmt.execute();

} catch (SQLException e) {
  System.err.println(e.getMessage());
}
```

#### Query Data

> In order to query data from YugabyteDB tables, execute the `SELECT` statement using `java.sql.statement` interface. Query results are returned using `java.sql.ResultSet` interface which can be iterated using `resultSet.next()` method for reading the data. Read more on [ResultSet](https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSet.html).

Reviewer: ```suggestion
To query data in YugabyteDB tables, execute the `SELECT` statement using the `java.sql.statement` interface. Query results are returned using the `java.sql.ResultSet` interface, which can be iterated using the `resultSet.next()` method for reading the data. Refer to [ResultSet](https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSet.html) in the Java documentation.
```

For example

```sql
SELECT * from employee;
```

```java
Connection conn = DriverManager.getConnection("jdbc:postgresql://localhost:5433/yugabyte","yugabyte", "yugabyte");
Statment stmt = conn.createStatement();
try {

  ResultSet rs = stmt.executeQuery("SELECT * FROM employee");
  while (rs.next()) {
    System.out.println(String.format("Query returned: name = %s, age = %s, language = %s", rs.getString(2), rs.getString(3), rs.getString(4)));
  }

} catch (SQLException e) {
  System.err.println(e.getMessage());
}
```

### Configure SSL/TLS

> In order to build a Java application that communicate securily over SSL with YugabyteDB database, get the root certificate (`ca.crt`) of the YugabyteDB Cluster. If certificates are not generated yet, follow the instructions in [Create server certificates](../../../../secure/tls-encryption/server-certificates/).

Reviewer: ```suggestion
To build a Java application that communicates securely over SSL, get the root certificate (`ca.crt`) of the YugabyteDB Cluster. If certificates are not generated yet, follow the instructions in [Create server certificates](../../../../secure/tls-encryption/server-certificates/).
```

> Generally when configuring the Java Client to use SSL, all the certs required for connecting to the database will be available on the Classpath of the Java application or default Postgres working directory `(~/.postgresql/)` of the VM or container where the app is hosted.

Reviewer: ```suggestion
Generally, when configuring the Java client to use SSL, all the certificates required for connecting to the database are available on the Classpath of the Java application, or in the default PostgreSQL working directory `(~/.postgresql/)` of the VM or container where the application is hosted.
```

---
title: JDBC Drivers
linkTitle: JDBC Drivers
description: JDBC Drivers for YSQL
headcontent: JDBC Drivers for YSQL
image: /images/section_icons/sample-data/s_s1-sampledata-3x.png
menu:
  latest:
    name: JDBC Drivers
    identifier: ref-postgres-jdbc-driver
    parent: drivers
    weight: 600
isTocNested: true
showAsideToc: true
---

<ul class="nav nav-tabs-alt nav-tabs-yb">

  <li >
    <a href="/latest/reference/drivers/java/yugabyte-jdbc-reference/" class="nav-link">
      <i class="icon-java-bold" aria-hidden="true"></i>
      YugabyteDB JDBC Driver
    </a>
  </li>

  <li >
    <a href="/latest/reference/drivers/java/postgres-jdbc-reference/" class="nav-link active">
      <i class="icon-postgres" aria-hidden="true"></i>
      PostgreSQL JDBC Driver
    </a>
  </li>

</ul>

The [PostgreSQL JDBC driver](https://jdbc.postgresql.org/) is the official JDBC driver for PostgreSQL which can used for connecting to YugabyteDB YSQL. YugabyteDB YSQL has full compatiblity with PostgreSQL JDBC Driver, allows Java programmers to connect to YugabyteDB database to execute DMLs and DDLs using the JDBC APIs.

> ## Quick Start

Author: I have updated the driver page to only have a CRUD example. Most of the low details are captured on this page.

JDBC Driver `java.sql.Connection` interface provides `connection.setAutoCommit()`, `connection.commit()` and `connection.rollback()` methods for enabling transactional access to YugabyteDB Database.

For example

```Java
Connection conn = DriverManager.getConnection("jdbc:postgresql://localhost:5433/yugabyte","yugabyte", "yugabyte");
conn.setAutoCommit(false);
Statment stmt = conn.createStatement();
try {
  
  PreparedStatement pstmt = connection.prepareStatement("INSERT INTO employees (id, name, age, langugage) VALUES (?, ?, ?, ?)");
  pstmt.setInt(1, 1);
  pstmt.setString(2, "John");
  pstmt.setInt(3, 35);
  pstmt.setString(4, "Java");
  pstmt.execute();

  conn.commit();

} catch (SQLException e) {
  System.err.println(e.getMessage());
}

```

By default PostgreSQL JDBC driver will have `auto-commit` mode enabled which means each SQL statement is treated as a transaction and is automatically committed. If one or more SQL statements are encapsulated in a transaction, auto-commit mode must be disabled by using `setAutoCommit()` method of the `Connection` object.

> ## Compatibility Matrix

Reviewer: redundant matrix

Author: removed.

| Driver Version | YugabyteDB Version | Support |
| :------------- | :----------------- | :------ |
| 42.2.14 | 2.11 (latest) | full
| 42.2.14 |  2.8 (stable) | full
| 42.2.14 | 2.6 | full

## Other Usage Examples

- [JDBC SSL example](../../../../quick-start/build-apps/java/ysql-jdbc-ssl/)

> - [ORM Examples]()

Reviewer: bad link

Author: I will resolve the link.

Author: Other Usage examples are not required here. We have removed it for now.

> ## FAQ

Reviewer: Empty section

Author: We will remove it and the FAQ section as required.
